File: src/csharp/oop/FileSearch/Search/SearchEngine.cs
using Microsoft.Extensions.FileProviders;

namespace FileSearch.Search;

public sealed class SearchEngine(IFileProvider fileProvider)
{
    public SearchEngine(string rootPath)
        : this(new PhysicalFileProvider(rootPath)) { }

    public IEnumerable<IFileInfo> Search(IMatchCriteria criteria, SearchOptions? searchOptions = null)
    {
        if (criteria is null)
        {
            return [];
        }
        
        searchOptions ??= new SearchOptions();

        var result = new List<IFileInfo>();

        var stack = new Stack<string>();
        stack.Push(Path.DirectorySeparatorChar.ToString());
        
        while (stack.Count > 0)
        {
            var currentPath = stack.Pop();
            var content = fileProvider.GetDirectoryContents(currentPath);

            foreach (var fileInfo in content)
            {
                if (!fileInfo.Exists)
                {
                    continue;
                }
                
                if (searchOptions.IsRecursive && fileInfo.IsDirectory)
                {
                    stack.Push(Path.Combine(currentPath, fileInfo.Name));
                }

                if (criteria.IsMatch(fileInfo, searchOptions))
                {
                    result.Add(fileInfo);
                }
            }
        }

        return result;
    }
}

File: src/python/oop/FileSearch/search/search_engine.py
import os
from typing import List
from search import *

class SearchEngine:
    def __init__(self, basedir: str):
        if not basedir:
            raise ValueError("The basedir cannot be empty")

        self.__basedir = basedir

    def search(self, criteria: MatchCriteriaBase, options: SearchOptions) -> List[FileInfo]:
        if not criteria:
            return []

        foundFiles = []

        stack = []
        stack.append(self.__basedir)

        while stack:
            dir = stack.pop()
            with os.scandir(dir) as entries:
                for entry in entries:
                    if entry.is_file():
                        file = FileInfo(entry)
                        if criteria.is_match(file, options):
                            foundFiles.append(file)
                    elif entry.is_dir() and options.is_recursive:
                        stack.append(entry.path)

        return foundFiles
